fdbserver/KeyValueStoreShardedRocksDB.actor.cpp

```diff
@@ -637,6 +775,61 @@ class ShardManager {
 		}
 	}
 
+	void persistRangeMapping(KeyRangeRef range, bool isAdd) {
+		TraceEvent(SevDebug, "ShardedRocksDB")
+		    .detail("Info", "RangeToPersist")
+		    .detail("BeginKey", range.begin)
+		    .detail("EndKey", range.end);
+		writeBatch->DeleteRange(metadataShard->cf,
+		                        getShardMappingKey(range.begin, shardMappingPrefix),
+		                        getShardMappingKey(range.end, shardMappingPrefix));
+
+		KeyRef lastKey = range.end;
+		if (isAdd) {
+			auto ranges = dataShardMap.intersectingRanges(range);
+			for (auto it = ranges.begin(); it != ranges.end(); ++it) {
+				if (it.value()) {
+					ASSERT(it.range() == it.value()->range);
+					// Non-empty range.
+					writeBatch->Put(metadataShard->cf,
+					                getShardMappingKey(it.range().begin, shardMappingPrefix),
+					                it.value()->physicalShard->id);
+					TraceEvent(SevDebug, "ShardedRocksDB")
+					    .detail("Action", "PersistRangeMapping")
+					    .detail("BeginKey", it.range().begin)
+					    .detail("EndKey", it.range().end)
+					    .detail("ShardId", it.value()->physicalShard->id);
+
+				} else {
+					// Empty range.
+					writeBatch->Put(metadataShard->cf, getShardMappingKey(it.range().begin, shardMappingPrefix), "");
```

Comment:
Do we need to set the beginKey of the overlapping range? It should have been set by a previous shard operation?

Also, if it.range().begin <range.end, will this value be incorrect, since we already `Put` the clearRange.